pystac/stac_io.py

```diff
@@ -1,61 +1,281 @@
-    to expand the ability of PySTAC to read different file systems.
-    For example, a client of the library might replace this class
-    member in it's own __init__.py with a method that can read from
-    cloud storage.
+
+class DuplicateObjectKeyError(Exception):
+    pass
+
+
+class DuplicateKeyReportingMixin(StacIO):
+    """A mixin for StacIO implementations that will report
+    on duplicate keys in the JSON being read in.
+
+    See https://github.com/stac-utils/pystac/issues/313
     """
 
-    write_text_method = default_write_text_method
-    """Users of PySTAC can replace the write_text_method in order
-    to expand the ability of PySTAC to write to different file systems.
-    For example, a client of the library might replace this class
-    member in it's own __init__.py with a method that can read from
-    cloud storage.
+    def _json_loads(self, txt: str, source: Union[str, "Link_Type"]) -> Dict[str, Any]:
+        return json.loads(
+            txt, object_pairs_hook=self.duplicate_object_names_report_builder(source)
+        )
+
+    @staticmethod
+    def duplicate_object_names_report_builder(
+        source: Union[str, "Link_Type"]
+    ) -> Callable[[List[Tuple[str, Any]]], Dict[str, Any]]:
+        def report_duplicate_object_names(
+            object_pairs: List[Tuple[str, Any]]
+        ) -> Dict[str, Any]:
+            result: Dict[str, Any] = {}
+            for key, value in object_pairs:
+                if key in result:
+                    url = (
+                        source
+                        if isinstance(source, str)
+                        else source.get_absolute_href()
+                    )
+                    raise DuplicateObjectKeyError(
+                        f"Found duplicate object name “{key}” in “{url}”"
+                    )
+                else:
+                    result[key] = value
+            return result
+
+        return report_duplicate_object_names
+
+
+class STAC_IO:
+    """DEPRECATED: Methods used to read and save STAC json.
+    Allows users of the library to set their own methods
+    (e.g. for reading and writing from cloud storage)
+
+    Note: The static methods of this class are deprecated. Move to using
+    instance methods of a specific instance of StacIO.
     """
 
-    # Replaced in __init__ to account for extension objects.
-    _stac_object_from_dict = None
+    @staticmethod
+    def read_text_method(uri: str) -> str:
+        return StacIO.default().read_text(uri)
+
+    @staticmethod
+    def write_text_method(uri: str, txt: str) -> None:
+        """Default method for writing text."""
+        return StacIO.default().write_text(uri, txt)
+
+    @staticmethod
+    def stac_object_from_dict(
+        d: Dict[str, Any],
+        href: Optional[str] = None,
+        root: Optional["Catalog_Type"] = None,
+    ) -> "STACObject_Type":
+        return pystac.serialization.stac_object_from_dict(d, href, root)
```

Comment:
Should we add  a `DeprecationWarning` to each of these in case folks don't read the docs?

Reply:
Good call - tbh I never figured out how to do deprecation properly in Python. This would be a matter of doing

```python
warnings.warn("STAC_IO is deprecated, yada yada", DeprecationWarning)
```

yeah?

Reply:
https://pypi.org/project/Deprecated/ doesn't seem worth adding an extra dependency